### PR TITLE
Implement `current_time` scalar function

### LINKED ISSUE #3982: Implement `current_time` Function
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As we add more full featured date/time support for DataFusion users need to be able to use the current time to calculate things like "all data in the last 5 hours" 

**Describe the solution you'd like**
Support the `current_time` postgres function https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT

Note I think we should avoid the `current_time(precision)` variant initially -- we can implement it as a follow on PR.

**Describe alternatives you've considered**

**Additional context**
This is a feature request we got from an early (internal) user of InfluxDB IOx

I believe this is a good first issue as you can follow the existing implementation of  `now()` 
- [ ] Physical support: https://github.com/apache/arrow-datafusion/blob/f1322592046a7ea895820286a2ad16d934540d92/datafusion/physical-expr/src/functions.rs#L424-L429) and use the same `query_execution_start_time` field
- [ ] Tests: https://github.com/apache/arrow-datafusion/blob/beeb631d0bb979e9b6826f6cd3eae2d0a24f471e/datafusion/core/tests/sql/timestamp.rs#L417

It would be a good exercise to add a new build in scalar function, physical planing support, and then tests

Note this the same ticket exists for the `current_date` function https://github.com/apache/arrow-datafusion/issues/3981

### PR DESCRIPTION
**Pull request by naosense**
# Which issue does this PR close?

Closes #3982 .

# Rationale for this change
DataFusion users need to be able to use the current time to calculate things like "all data in the last 5 hours"

# What changes are included in this PR?
Scalar `current_time` function

# Are there any user-facing changes?

### CODE REVIEW

**Comment by waitingkuo** [comment]
LGTM, thank you @naosense

**Comment by naosense** [comment]
I followed @comphead 's `current_date` test case
```
let sql = "select case when current_time() = cast(now() as time) then 'OK' else 'FAIL' end result";
    let results = execute_to_batches(&ctx, sql).await;

    let expected = vec![
        "+--------+",
        "| result |",
        "+--------+",
        "| OK     |",
        "+--------+",
    ];

    assert_batches_eq!(expected, &results);
```
but it reported an error
```
thread 'sql::timestamp::test_current_time' panicked at 'called `Result::unwrap()` on an `Err` value: "NotImplemented(\"Unsupported CAST from Timestamp(Nanosecond, Some(\\\"UTC\\\")) to Time64(Nanosecond)\") at Creating physical plan for 'select case when current_time() = cast(now() as time) then 'OK' else 'FAIL' end result': Projection: CASE WHEN currenttime() = CAST(now() AS Time64(Nanosecond)) THEN Utf8(\"OK\") ELSE Utf8(\"FAIL\") END AS result\n  EmptyRelation"', datafusion/core/tests/sql/mod.rs:806:10
```
Is anything i missed?

**Comment by mingmwang** [comment]
If there are multiple current_time() in a SQL statement, will it be invoked multiple times or just once ?

**Comment by waitingkuo** [comment]
casting `Timestamp` to `Time` isn't supported in `arrow-rs` 
https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/cast.rs#L236-L265

**Comment by waitingkuo** [comment]
@naosense 
you can do the following for now if you want to add that test case
```bash
❯ select (now()::bigint % 86400000000000)::time;
+-------------------------------+
| now() % Int64(86400000000000) |
+-------------------------------+
| 07:41:27.421667               |
+-------------------------------+
```

i'll be great if you could fire a issue (or even a pr) to arrow-rs

File: datafusion/physical-expr/src/functions.rs [hunk 1]
```diff
@@ -433,6 +433,12 @@ pub fn create_physical_fun(
                 execution_props.query_execution_start_time,
             ))
         }
+        BuiltinScalarFunction::CurrentTime => {
+            // bind value for current_time at plan time
+            Arc::new(datetime_expressions::make_current_time(
+                execution_props.query_execution_start_time,
+            ))
```
Comment (waitingkuo) [inline datafusion/physical-expr/src/functions.rs:440]: 👍 

@mingmwang all the current_time within a query uses the same query_execution_start_time

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 2]
```diff
@@ -200,6 +200,20 @@ pub fn make_current_date(
...
+/// specified current time.
+///
+/// The semantics of `current_time()` require it to return the same value
+/// wherever it appears within a single statement. This value is
+/// chosen during planning time.
+pub fn make_current_time(
+    now_ts: DateTime<Utc>,
+) -> impl Fn(&[ColumnarValue]) -> Result<ColumnarValue> {
+    let nano =
+        Duration::seconds(now_ts.num_seconds_from_midnight() as i64).num_nanoseconds();
```
Comment (naosense) [inline datafusion/physical-expr/src/datetime_expressions.rs:213]: I found a bug, nanoseconds less than a second will be loss if `num_seconds_from_midnight()` is used,     I will do  this in another way

Comment (waitingkuo) [reply datafusion/physical-expr/src/datetime_expressions.rs]: i think you might not need Duration::seconds here.

```rust
let nano = (now_ts.num_seconds_from_midnight() as i64) * 1000000000 + (now_ts.nanosecond() as i64);
```

 [1] [nanosecond](https://docs.rs/chrono/latest/chrono/trait.Timelike.html#tymethod.nanosecond)

**Review by waitingkuo — APPROVED** [review]
thank you @naosense

File: datafusion/core/tests/sql/timestamp.rs [hunk 3]
```diff
@@ -1651,3 +1651,34 @@ async fn test_current_date() -> Result<()> {
...
+    let sql = "select case when current_time() = (now()::bigint % 86400000000000)::time then 'OK' else 'FAIL' end result";
+    let results = execute_to_batches(&ctx, sql).await;
+
+    let expected = vec![
+        "+--------+",
+        "| result |",
+        "+--------+",
+        "| OK     |",
+        "+--------+",
+    ];
```
Comment (waitingkuo) [inline datafusion/core/tests/sql/timestamp.rs:1680]: 👍

**Comment by naosense** [comment]
> @waitingkuo 
> 
> i'll be great if you could fire a issue (or even a pr) to arrow-rs to support  cast Timestamp to Time

it's a pleasure, I'll investigate how to do that

**Comment by naosense** [comment]
> @waitingkuo 
> 
> i'll be great if you could fire a issue (or even a pr) to arrow-rs to support  cast Timestamp to Time

it's a pleasure, I'll investigate how to do that

**Review by alamb — APPROVED** [review]
This looks great @naosense . Thanks @waitingkuo  for the review!

**Comment by alamb** [comment]
Hi @naosense  -- it seems this PR needs to have `cargo fmt` to run on it in order to pass CI. I took the liberty of doing so (and merging up from master) to prevent another 24 hour cycle

**Comment by alamb** [comment]
Thanks again @naosense  and @waitingkuo !

**Pull request merged**